lightning/src/ln/channelmanager.rs

```diff
@@ -8296,6 +8296,57 @@ where
 						}
 					}
 				}
+
+				// Whether the downstream channel was closed or not, try to re-apply any payment
+				// preimages from it which may be needed in upstream channels for forwarded
+				// payments.
+				let outbound_claimed_htlcs_iter = monitor.get_all_current_outbound_htlcs()
+					.into_iter()
+					.filter_map(|(htlc_source, (htlc, preimage_opt))| {
+						if let HTLCSource::PreviousHopData(prev_hop_data) = htlc_source {
+							if let Some(payment_preimage) = preimage_opt {
+								Some((prev_hop_data, htlc, payment_preimage))
+							} else { None }
+						} else { None }
+					});
+				for (prev_hop_data, htlc, payment_preimage) in outbound_claimed_htlcs_iter {
+					let mut is_chan_open = false;
+					if let Some((node_id, chan_id)) = short_to_chan_info.get(&prev_hop_data.short_channel_id) {
+						if let Some(mut peer) = per_peer_state.get_mut(node_id).map(|node| node.lock().unwrap()) {
+							if let Some(chan) = peer.channel_by_id.get_mut(chan_id) {
+								is_chan_open = true;
+								match chan.get_update_fulfill_htlc_and_commit(prev_hop_data.htlc_id, payment_preimage, &args.logger) {
+									UpdateFulfillCommitFetch::DuplicateClaim {} => {},
+									UpdateFulfillCommitFetch::NewClaim { monitor_update, .. } => {
+										// The ChannelMonitor that gave us this
+										// preimage is for a now-closed channel -
+										// no further updates to that channel can
+										// happen which would result in the
+										// preimage being removed, thus we're
+										// guaranteed to regenerate this claim on
+										// restart as long as the source monitor
```

Comment:
How do we know that the downstream channel is closed now? If it is, then the comment on the `iter` above needs updating. 

Also `s/source monitor/downstream monitor`? "source" makes me think it's the upstream monitor, though I think that might mean "source of the preimage."